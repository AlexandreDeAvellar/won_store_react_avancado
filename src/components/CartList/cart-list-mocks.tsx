import { CartListProps } from '.'

export const cartListProps: CartListProps = {}
