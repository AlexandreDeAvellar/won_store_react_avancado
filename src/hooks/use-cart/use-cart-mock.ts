import { graphql_host } from '../../utils/graphql-transform/graphql_host'
import { QUERY_GAMES } from '../../graphql/queries/games'

export const mockUseCartItems = {
  data: [
    {
      attributes: {
        name: 'Enter the Gungeon',
        slug: 'enter-the-gungeon',
        cover: {
          data: {
            attributes: {
              url: '/uploads/enter_the_gungeon_46f4fe4c68.jpg'
            }
          }
        },
        developers: {
          data: [
            {
              attributes: {
                name: 'Dodge Roll'
              }
            }
          ]
        },
        price: 27.99,
        __typename: 'Game'
      },
      id: '1'
    },
    {
      attributes: {
        name: 'The Witcher: Enhanced Edition',
        slug: 'the-witcher',
        cover: {
          data: {
            attributes: {
              url: '/uploads/the_witcher_fe844be564.jpg'
            }
          }
        },
        developers: {
          data: [
            {
              attributes: {
                name: 'CD PROJEKT RED'
              }
            }
          ]
        },
        price: 10.5,
        __typename: 'Game'
      },
      id: '2'
    }
  ]
}

export const useCartGamesMock = {
  request: {
    query: QUERY_GAMES,
    variables: { limit: 1000, start: 0, where: { id: { in: ['1', '2'] } } }
  },
  result: {
    data: {
      gamesConnection: {
        meta: {
          pagination: {
            total: 2
          }
        }
      },
      games: mockUseCartItems
    }
  }
}

export const cartItems = [
  {
    id: '1',
    img: graphql_host + '/uploads/enter_the_gungeon_46f4fe4c68.jpg',
    price: '$27.99',
    title: 'Enter the Gungeon'
  },
  {
    id: '2',
    img: graphql_host + '/uploads/the_witcher_fe844be564.jpg',
    price: '$10.50',
    title: 'The Witcher: Enhanced Edition'
  }
]
