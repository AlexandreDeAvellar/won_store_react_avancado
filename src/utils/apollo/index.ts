import { ApolloClient, HttpLink, NormalizedCacheObject } from '@apollo/client'
import { setContext } from '@apollo/client/link/context'
import { useMemo } from 'react'
import { apolloCache } from './apollo-cache'
import { Session } from 'next-auth'

let apolloClient: ApolloClient<NormalizedCacheObject> | null

const createApolloClient = (session?: Session | null) => {
  // https://www.apollographql.com/docs/react/networking/authentication/

  const httpLink = new HttpLink({ uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql` })
  const authLink = setContext((_, { headers, session: clientSession }) => {
    const jwt = session?.jwt || clientSession?.jwt || ''
    const authorization = jwt ? `Bearer ${session?.jwt}` : ''
    return {
      headers: { ...headers, authorization }
    }
  })
  return new ApolloClient({
    link: authLink.concat(httpLink),
    ssrMode: typeof window === 'undefined',
    cache: apolloCache
  })
}

export const initializeApollo = (initialState = null, session?: Session | null) => {
  // serve para verificar se já existe uma instância, para não criar outra
  const apoloClientGlobal = apolloClient ?? createApolloClient(session)

  // se a página usar o apolloClient no lado client
  // hidratamos o estado inicial aqui
  if (initialState) {
    apoloClientGlobal.cache.restore(initialState)
  }

  // sempre inicializando no SSR com cache limpo
  if (typeof window === 'undefined') return apoloClientGlobal

  // cria o apolloClient se estiver no client side
  apolloClient = apolloClient ?? apoloClientGlobal

  return apolloClient
}

export const useApollo = (initialState = null, session?: Session | null) => {
  const store = useMemo(() => initializeApollo(initialState, session), [initialState, session])
  return store
}
