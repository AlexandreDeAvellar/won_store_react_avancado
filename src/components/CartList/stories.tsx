import { StoryObj, Meta } from '@storybook/react'
import CartList, { CartListProps } from '.'

export default {
  title: 'CartList',
  component: CartList,
  argsTypes: {
    cartContextValue: { type: '' },
    items: { type: '' }
  }
} as Meta

type Story = StoryObj<typeof CartList>

export const Default: Story = (args: CartListProps) => (
  <div style={{ maxWidth: 800 }}>
    <CartList {...args} />
  </div>
)

export const WithButton: Story = (args: CartListProps) => (
  <div style={{ maxWidth: 800 }}>
    <CartList {...args} hasButton />
  </div>
)

export const Empty = () => (
  <div style={{ maxWidth: 800 }}>
    <CartList />
  </div>
)

Default.args = { hasButton: false }
WithButton.args = { hasButton: true }
Empty.args = { cartContextValue: {} }
