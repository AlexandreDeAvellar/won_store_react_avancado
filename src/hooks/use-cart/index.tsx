import { createContext, useContext, useEffect, useState } from 'react'
import { useQueryGames } from '../../graphql/queries/games'

import { getStorageItem } from '../../utils/localStorage'
import { currencyFormat } from '../../utils/format'
import { cartTransform } from '../../utils/graphql-transform'

const CART_KEYS = 'cartItems'

export type CartItem = {
  id: string
  img: string
  title: string
  price: string
}

export type CartContextProps = {
  items: CartItem[]
  quantity: number
  total: string
}

export const defaultCartItem: CartContextProps = { items: [], quantity: 0, total: '$0.00' }

export const CartContext = createContext<CartContextProps>(defaultCartItem)

export type CartProviderProps = {
  children: React.ReactNode
}

export const CartProvider = ({ children }: CartProviderProps) => {
  const [cartItems, setCartItems] = useState<string[]>([])

  useEffect(() => {
    const idGames = getStorageItem(CART_KEYS)
    if (idGames) setCartItems(idGames)
  }, [])

  const { data } = useQueryGames({
    limit: 1000,
    start: 0,
    filters: { where: { id: { in: cartItems } } }
  })

  const items = cartTransform(data?.games)
  const quantity = cartItems.length
  const total = data?.games?.data?.reduce((acc: number, game: { attributes: { price: number } }) => {
    return acc + game.attributes.price
  }, 0)

  return <CartContext.Provider value={{ items, quantity, total: currencyFormat(total || 0) }}>{children}</CartContext.Provider>
}

export const useCart = () => useContext(CartContext)
