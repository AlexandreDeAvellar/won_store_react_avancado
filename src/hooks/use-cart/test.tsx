import { renderHook, waitFor } from '@testing-library/react'
import { CartProvider, CartProviderProps, useCart } from '.'
import { setStorageItem } from '../../utils/localStorage'
import { cartItems, useCartGamesMock } from './use-cart-mock'
import { MockedProvider } from '@apollo/client/testing'

describe('useCart', () => {
  const ids = ['1', '2']

  it('should render items and its info if there are any in the cart', async () => {
    const wrapper = ({ children }: CartProviderProps) => (
      <MockedProvider mocks={[useCartGamesMock]}>
        <CartProvider>{children}</CartProvider>
      </MockedProvider>
    )

    setStorageItem('cartItems', ids)
    const { result } = renderHook(() => useCart(), { wrapper })

    await waitFor(() => expect(result.current.items).toStrictEqual(cartItems))
    await waitFor(() => expect(result.current.quantity).toBe(2))
    await waitFor(() => expect(result.current.total).toBe('$38.49'))
  })
})
