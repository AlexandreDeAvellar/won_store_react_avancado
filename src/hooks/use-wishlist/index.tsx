import { createContext, useContext, useEffect, useMemo, useState } from 'react'
import { useSession } from 'next-auth/react'
import { useMutation } from '@apollo/client'

import { GameCardProps } from '../../components/GameCard'
import { useQueryWishlist } from '../../graphql/queries/wishlist'
import { GameCardGraphqlProps, gameCardTransform } from '../../utils/graphql-transform'
import {
  MUTATION_CREATE_WISHLIST,
  MUTATION_UPDATE_WISHLIST,
  MutationCreateWishlistProps,
  MutationCreateWishlistVariables,
  MutationUpdateWishlistProps,
  MutationUpdateWishlistVariables
} from '../../graphql/mutations/wishlist'

export type WishlistContextProps = {
  items: GameCardProps[]
  isInWishlist: (id: string) => boolean
  addWishlist: (id: string) => void
  removeFromWishlist: (id: string) => void
  loading: boolean
}

export const WishlistContextDefault: WishlistContextProps = {
  items: [],
  isInWishlist: () => false,
  addWishlist: () => null,
  removeFromWishlist: () => null,
  loading: false
}

export const WishlistContext = createContext<WishlistContextProps>(WishlistContextDefault)

export type WishlistContextProviderProps = { children: React.ReactNode }

export const WishlistProvider = ({ children }: WishlistContextProviderProps) => {
  const { data: session } = useSession()
  const [wishlistItems, setWishlistItems] = useState<GameCardGraphqlProps>()
  const [wishListId, setWishlistId] = useState<string | null>()

  const { data, loading } = useQueryWishlist({
    skip: !session?.user?.email,
    context: { session },
    variables: { identifier: session?.user?.email as string }
  })

  useEffect(() => {
    setWishlistItems(data?.wishlists?.data[0]?.attributes?.games)
    setWishlistId(data?.wishlists?.data[0]?.id)
  }, [data])

  const [createWishlist] = useMutation<MutationCreateWishlistProps, MutationCreateWishlistVariables>(MUTATION_CREATE_WISHLIST, {
    context: { session },
    onCompleted: ({ createWishlist: { data } }) => {
      setWishlistItems(data?.attributes?.games)
      setWishlistId(data?.id)
    }
  })

  const [updateWishlist] = useMutation<MutationUpdateWishlistProps, MutationUpdateWishlistVariables>(MUTATION_UPDATE_WISHLIST, {
    context: { session },
    onCompleted: ({ updateWishlist: { data } }) => {
      setWishlistItems(data?.attributes?.games)
    }
  })

  const isInWishlist = (id: string) => !!wishlistItems?.data.find((game) => game.id === id)

  const wishlistIds: string[] = useMemo(() => wishlistItems?.data.map((game) => game.id), [wishlistItems]) || []

  const addWishlist = (id: string) => {
    if (!session?.user) return
    if (!wishListId) return createWishlist({ variables: { user: session?.user?.id, games: [...wishlistIds, id] } })

    updateWishlist({ variables: { id: wishListId, data: { user: session?.user?.id, games: [...wishlistIds, id] } } })
  }

  const removeFromWishlist = (id: string) => {
    if (!session?.user) return
    const gameId = wishlistIds.filter((gameId) => gameId !== id)
    if (wishListId) return updateWishlist({ variables: { id: wishListId, data: { user: session?.user?.id, games: gameId } } })
  }

  let items: GameCardProps[] = []

  if (!loading && wishlistItems) {
    items = wishlistItems ? gameCardTransform(wishlistItems) : []
  }

  const values = { items, isInWishlist, addWishlist, removeFromWishlist, loading }
  return <WishlistContext.Provider value={values}>{children}</WishlistContext.Provider>
}

export const useWishlist = () => useContext(WishlistContext)
