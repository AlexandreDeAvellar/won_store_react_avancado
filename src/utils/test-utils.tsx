import { ReactNode } from 'react'

import { CartContext, CartContextProps, defaultCartItem } from '../hooks/use-cart'
import { RenderOptions, render } from '@testing-library/react'

import { ThemeProvider } from 'styled-components'
import theme from '../styles/theme'

export type CustomRenderProps = {
  cartProviderProps?: CartContextProps
} & Omit<RenderOptions, 'queries'>

export const customRender = (ui: ReactNode, { cartProviderProps = defaultCartItem, ...renderOptions }: CustomRenderProps = {}) =>
  render(
    <ThemeProvider theme={theme}>
      <CartContext.Provider value={cartProviderProps}>{ui}</CartContext.Provider>
    </ThemeProvider>,
    renderOptions
  )

export * from '@testing-library/react'
export { customRender as render }
