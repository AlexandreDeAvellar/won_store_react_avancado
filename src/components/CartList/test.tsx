import { render as renderWithTheme, screen } from '../../utils/test-utils'

import CartList from '.'
import { cartItems } from '../../hooks/use-cart/use-cart-mock'
import { defaultCartItem } from '../../hooks/use-cart'

describe('<CartList />', () => {
  it('should render the cart list', () => {
    const { container } = renderWithTheme(<CartList />, { cartProviderProps: { ...defaultCartItem, items: cartItems, total: 'R$ 38,49' } })

    expect(screen.getAllByRole('heading')).toHaveLength(2)
    expect(screen.getByText('R$ 38,49')).toHaveStyle({ color: '#F231A5' })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the button', () => {
    renderWithTheme(<CartList hasButton />, { cartProviderProps: { ...defaultCartItem, items: cartItems } })

    expect(screen.getByText(/buy it now/i)).toBeInTheDocument()
  })

  it('should render empty if there are no games', () => {
    renderWithTheme(<CartList />)

    expect(screen.getByText(/your cart is empty/i)).toBeInTheDocument()
    expect(screen.queryByText(/total/i)).not.toBeInTheDocument()
  })
})
