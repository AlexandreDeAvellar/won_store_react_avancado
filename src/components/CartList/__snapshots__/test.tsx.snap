// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartList /> should render the cart list 1`] = `
.c2 {
  padding: 1.2rem;
  border-bottom: 0.1rem solid #EAEAEA;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 1.2rem;
  width: 9.6rem;
  height: 5.6rem;
}

.c4 img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  width: 100%;
}

.c6 {
  font-size: 0.8rem;
  line-height: 0.8rem;
  font-weight: 600;
  color: #030517;
  margin-bottom: 0.4rem;
}

.c8 {
  color: #FAFAFA;
  padding: 0.2rem 0.4rem;
  background-color: #3CD3C1;
  border-radius: 0.4rem;
  font-size: 0.8rem;
  font-weight: 600;
}

.c0 {
  background: #FAFAFA;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-self: start;
  -ms-flex-item-align: start;
  align-self: start;
}

.c1 {
  max-height: 40rem;
  overflow-y: auto;
}

.c9 {
  background: #eee;
  color: #030517;
  font-weight: 600;
  font-size: 0.8rem;
  padding: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c10 {
  color: #F231A5;
}

@media (min-width:768px) {
  .c2 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media (min-width:768px) {
  .c4 {
    margin-right: 0.8rem;
    width: 15rem;
    height: 7rem;
  }
}

@media (min-width:768px) {
  .c6 {
    font-size: 2rem;
    line-height: 2rem;
  }
}

@media (min-width:768px) {
  .c9 {
    font-size: 1rem;
    padding: 1.2rem;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <img
            alt="Enter the Gungeon"
            src="http://127.0.0.1:1337/uploads/enter_the_gungeon_46f4fe4c68.jpg"
          />
        </div>
        <div
          class="c5"
        >
          <h3
            class="c6"
          >
            Enter the Gungeon
          </h3>
          <div
            class="c7"
          >
            <div
              class="c8"
            >
              $27.99
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <img
            alt="The Witcher: Enhanced Edition"
            src="http://127.0.0.1:1337/uploads/the_witcher_fe844be564.jpg"
          />
        </div>
        <div
          class="c5"
        >
          <h3
            class="c6"
          >
            The Witcher: Enhanced Edition
          </h3>
          <div
            class="c7"
          >
            <div
              class="c8"
            >
              $10.50
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c9"
  >
    <span>
      Total:
    </span>
    <span
      class="c10"
    >
      R$ 38,49
    </span>
  </div>
</div>
`;
